### PR TITLE
167 - Upload of branding and upload of partners

### PR DESCRIPTION
**Pull request by latin-panda**
# Description
This PR will add two new commands:

1. `upload-branding` will update the medic DB's branding document with: Title, resources (logo and favicon) and attachments.
2. `upload-partners` will update the medic DB's partners document with: resources (parner name and partner image) and attachments.

Additionally will add support to include as attachment the files with extension: ".icon"

https://github.com/medic/medic-conf/issues/167

# Code review items

- Readable: Concise, well named, follows the [style guide](https://github.com/medic/medic-docs/blob/master/development/style-guide.md), documented if necessary.
- Documented: Configuration and user documentation on [medic-docs](https://github.com/medic/medic-docs/)
- Tested: Unit and/or e2e where appropriate
- Internationalised: All user facing text
- Backwards compatible: Works with existing data and configuration or includes a migration. Any breaking changes documented in Changes.md.

# License

The software is provided under AGPL-3.0. Contributions to this project are accepted under the same license.

### CODE REVIEW

**Comment by latin-panda** [comment]
Hi @garethbowen, since branding and partners are separate documents in the DB, I thought on making two separate commands as well.
`upload-branding`
`upload-partners`

**Comment by MaxDiz** [comment]
If you can also update [this documentation](https://docs.communityhealthtoolkit.org/apps/reference/resources/) upon completion of this ticket, it would be appreciated!

File: src/fn/upload-partners.js [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+      info('Partners file uploaded');
+    } else {
+      info('Partners file not uploaded as no changes found');
+    }
+
+    warnUploadOverwrite.postUploadDoc(doc);
+
+    return Promise.resolve();
+  }
+};
```
Comment (garethbowen) [inline src/fn/upload-partners.js:42]: This is so similar to `upload-branding` and `upload-resources` there should be some way to make a lib function which all three can call to encourage reuse. At least lines 27-40 can be extracted, but I think if you may be able to pull out a function which takes the JSON path, the attachments path, and the base doc as parameters so this file just calls that function and then logs the outcome.

File: src/lib/attachment-from-file.js [hunk 2]
```diff
@@ -16,6 +16,7 @@ function mimeTypeFor(fileName) {
     case 'json': return 'application/json';
     case 'png' : return 'image/png';
     case 'svg' : return 'image/svg+xml';
+    case 'ico' : return 'image/x-icon';
```
Comment (garethbowen) [inline src/lib/attachment-from-file.js:19]: This section is greatly improved in this not yet merged branch: https://github.com/medic/medic-conf/pull/336/files

Double check that the new way works with `ico` attachments and then revert this change so it doesn't conflict.

File: test/fn/upload-partners.spec.js [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+      await uploadPartners.execute();
+
+      expect(attachmentsFromDir.called).to.be.true;
+      expect(pouch.called).to.be.true;
+      expect(warnUploadOverwrite.preUploadDoc.args[0][1]).to.deep.include(partnersDoc);
+      expect(warnUploadOverwrite.postUploadDoc.args[0][0]).to.deep.include(partnersDoc);
+      expect(insertOrReplace.args[0][1]).to.deep.include(partnersDoc);
+    });
+  });
+});
```
Comment (garethbowen) [inline test/fn/upload-partners.spec.js:74]: Also test the other conditions, like the directory doesn't exist, or no changes are required.

File: src/fn/upload-partners.js (hunk 1 above)
Comment (latin-panda) [reply src/fn/upload-partners.js:12]: Done!

File: src/lib/attachment-from-file.js (hunk 2 above)
Comment (latin-panda) [reply src/lib/attachment-from-file.js]: Done!

File: test/fn/upload-partners.spec.js (hunk 3 above)
Comment (latin-panda) [reply test/fn/upload-partners.spec.js:26]: Done!

File: .eslintrc [hunk 4]
```diff
@@ -5,7 +5,7 @@
   },
   "extends": "eslint:recommended",
   "parserOptions": {
-    "ecmaVersion": 8
+    "ecmaVersion": 9
```
Comment (latin-panda) [inline .eslintrc:8]: @garethbowen Hi, can I do this to support spread operator in the lint?

File: src/lib/main.js [hunk 5]
```diff
@@ -37,6 +39,8 @@ const defaultArchiveActions = [
   'upload-collect-forms',
   'upload-contact-forms',
   'upload-resources',
+  'upload-branding',
```
Comment (latin-panda) [inline src/lib/main.js:42]: @garethbowen Does it make sense to add the new commands here too? 🤔

File: src/lib/attachment-from-file.js (hunk 2 above)
Comment (latin-panda) [reply src/lib/attachment-from-file.js]: Fixed!

File: .eslintrc (hunk 4 above)
Comment (garethbowen) [reply .eslintrc:8]: We support Node 8 which does support the spread operator but it's really hard to prove if it supports _everything_ in ecma 9 so this change risks our linter not catching regressions. I think stick with version 8 for now - we will bump our support matrix to Node 12 (or 14 maybe?) soon so we'll be able to increase this then.

File: src/fn/upload-partners.js (hunk 1 above)
Comment (garethbowen) [reply src/fn/upload-partners.js:12]: Awesome!

File: src/lib/main.js (hunk 5 above)
Comment (garethbowen) [reply src/lib/main.js:42]: Yes! Good call.

File: .eslintrc (hunk 4 above)
Comment (latin-panda) [reply .eslintrc]: Fixed!

**Comment by garethbowen** [comment]
Actually, I just noticed reviewing the documentation that you changed the format of the `resouces.json` file. I don't think we can do this because it would be a breaking change for everyone who's using it. Is this correct? Can you find a way to do it without breaking backwards compatibility?

**Comment by latin-panda** [comment]
Hi @garethbowen I could add any of the following options, but it would add some ugly noise in the code that I'd like to avoid if it's possible to announce the breaking change (maybe in conjunction with a node upgrade 🙂).

Option 1:
```
if command === upload_resources then 
      add json content inside doc.resources
else 
      extend doc with json content (my solution)
```

Option 2:
A dynamic option is to pass as parameter a "mapping object", so to instruct our generic function where to take the data, this will restrict what we are copying from the json (advantage or disadvantage 🤔)  for example:
Branding mapping
```
{
  title: 'title', // equivalent json.title
  icons: 'images.icon', // json.images.icon
}
```
Resources mapping
```
{
  otherProp: 'otherProp' // equivalent json.otherProp
  resources: '' // equivalent assigning the whole json
}
```

**Comment by garethbowen** [comment]
I'm strongly against breaking backwards compatibility. Even if we announce it there's a chance someone will miss it, and even if they don't it's annoying for every app builder!

Option 3 would be for each implementation to pass in the doc JSON rather than the path to the JSON so each caller has to parse the file but they can pass a subset of the file if they like.

Option 4 would be to pass an optional mapping function to turn the JSON file into the expected CouchDB doc, eg:

```js
return uploadConfigurationDocs(configurationPath, directoryPath, 'resources', file => file.resources);
```

By default no mapping would be applied.

I like Option 4 - what do you think?

**Comment by latin-panda** [comment]
Hi @garethbowen  I proceeded with option 4, added unit test and also manually tested that resources is working as before.

**Comment by garethbowen** [comment]
Thanks! Looks good.
